### PR TITLE
Defers output cache key creation until we know we need it

### PR DESCRIPTION
**Pull request by paynecrl97**
This will increase performance by a very slight amount (~1ms) on an OOTB install of Orchard. If someone has overridden the virtual `ComputeCacheKey` with more complex logic then the performance gains will be noticeable.

See #5390

### CODE REVIEW

File: src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs [hunk 1]
```diff
@@ -267,45 +268,47 @@ public class OutputCacheFilter : FilterProvider, IActionFilter, IResultFilter {
 
         protected virtual bool RequestIsCacheable(ActionExecutingContext filterContext) {
 
+            var itemDescriptor = string.Format("{0} (Controller: {1}, Action: {2})", filterContext.RequestContext.HttpContext.Request.RawUrl, filterContext.ActionDescriptor.ControllerDescriptor.ControllerName, filterContext.ActionDescriptor.ActionName);
```
Comment (paynecrl97) [inline src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs]: We previously logged the cache key as an identifier, but that will always be null now, so `itemDescriptor` is designed to be a lightweight replacement for that.

Comment (Piedone) [inline src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs]: Can you make sure that itemDescriptor really contains all debugging-relevant information? Just controller and action doesn't look enough. I'm thinking about at least query string parameters as well. However those can contain sensitive information...

Comment (paynecrl97) [inline src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs]: `{0}` is the raw URL, which will include the query string, eg: `/OrchardLocal?hello=1 (Controller: Item, Action: Display)`.

This value has already been logged at https://github.com/paynecrl97/Orchard/blob/5390/delay_output_cache_key_creation/src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs#L83, so I made the assumption that I was safe to log the full query string.

I could probably add in Tenant, Headers, Theme, Culture & Auth though (I *think* that those are all the params that are included in the OOTB cache key).

Comment (Piedone) [inline src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs]: Ah sorry, misread. I'd add Area and the other data you mentioned except tenant (that's already logged).

Comment (paynecrl97) [inline src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs]: Added those values, except for Headers as there is a bit of logic around which headers should be used to differentiate and I wanted to keep this as lightweight as possible.

File: src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs [hunk 2]
```diff
@@ -267,45 +268,55 @@ public class OutputCacheFilter : FilterProvider, IActionFilter, IResultFilter {
 
         protected virtual bool RequestIsCacheable(ActionExecutingContext filterContext) {
 
+            var url = filterContext.RequestContext.HttpContext.Request.RawUrl;
```
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs]: Please wrap this code by testing if `Logger.IsEnabled(LogLevel.Debug)` is true, then build `itemDescriptor`.

**Pull request merged**